grpc-for-smart-contracts.md

```diff
@@ -0,0 +1,263 @@
+    ```
+
+- Returns `null` if the address could not be found.
+
+## `GetInstances`
+
+*Retrieve the addresses of all smart contract instances.*
+
+```protobuf
+rpc GetInstances(BlockHash) returns (JsonResponse) {}
+
+message BlockHash {
+  string block_hash = 1;
+}
+```
+**Input:**
+
+- `block_hash`: A null-terminated base16 encoding of a block hash. 
+
+**Output:**
+
+- A JSON list of contract addresses on the chain.
+  - Example: `[{"subindex":0,"index":0},{"subindex":0,"index":1}]` 
+
+## `GetInstanceInfo`
+
+*Retrieve information about a given smart contract instance.*
+
+```protobuf
+rpc GetInstanceInfo(GetAddressInfoRequest) returns (JsonResponse) {}
+
+message GetAddressInfoRequest {
+  string block_hash = 1;
+  string address = 2;
+}
+```
+
+**Input:**
+
+- `block_hash`: A null-terminated base16 encoding of a block hash. 
+- `address`: A null-terminated JSON-encoded value (same format as returned by
+  `GetInstances`).
+
+**Output:**
+
+- A JSON object with information about the contract instance.
+  - Example: 
+
+  ```json
+  {"amount":"0","sourceModule":"609af98562314a5461990080df5514703bfb9f98d2c5b4535cf4b11f766cac29","owner":"4NyqjcJwKAckGE3gMGDJeqcyLZQxtfnDFTdvg74xYUU8myQTrB","methods":["counter.receive","counter.receive_optimized"],"name":"init_counter","model":"0000000000"}
+  ```
+
+
+## `SendTransaction`
+
+*Send a transaction.*
+
+```protobuf
+rpc SendTransaction(SendTransactionRequest) returns (BoolResponse) {}
+
+message SendTransactionRequest {
+  uint32 network_id = 1;
+  bytes payload = 2;
+}
+```
+
+**Input:**
+
+- `network_id`: Id for the network. Default id is `100`.
+- `payload`: Binary encoding of the transaction (details below).
+
+**Output:**
+
+- Boolean value indicating the transaction success.
+
+
+What follows is the description of the data types, and their serialization, which is
+needed to construct the `payload` in `SendTransaction`.
+It starts with a `BlockItem` which is the type expected as `payload`.
+
+### `BlockItem`
+
+A `BlockItem` is a union type with a version number.
+For interacting with smart contracts, the relevant variant is `AccountTransaction`.
+    
+#### Serialization
+
+- Version (uint32, big-endian)
```

Comment:
This is not correct, version encoding is variable-length, it is big-endian base  128 encoding.